### ISSUE TITLE
Is it possible to store shapes data as array-like objects?

### ISSUE THREAD

**Issue by hanjinliu** [napari/napari#4716]
## 🚀 Feature

It seems that `Shapes` layer convert all the input data into `np.ndarray`.
I wonder if it is possible to use custom class for shapes.

## Motivation

1. I usually use B-spline as a "ROI". I can use path shapes to visualize spline curves but coordinates of a path is not a true representation of a spline (just like `[np.arange(100), np.sin(np.arange(100)]` does not accurately represents the sine function). Of course I can keep the spline curves elsewhere but if `Shapes` layer can store them as is, it will be much simpler.

2. When a layer is composed of different types of shapes, I sometimes want to use custom arrays (like "Rectangle" or "Line"). The repr of data would be clearer:

  ```python
  layer.data
  ```
  ```
  [<Rectangle(...)>, <Line(...)>]
  ```

  ... and I can also switch methods without referring to `shape_type`.

  ```python
  # measure areas for every shape
  results = [shape.area() for shape in layer.data]  # suppose all the array objects have "area" method.
  ```

## Pitch

Accept any array-like objects that follows certain protocol, like `Image` layer does.

## Alternatives

Basically I can make it by keeping some additional information elsewhere and connecting data-changed event to it.
